src/miner.cpp

```diff
@@ -92,9 +93,9 @@ static BlockAssembler::Options DefaultOptions()
     return options;
 }
 
-BlockAssembler::BlockAssembler(CChainState& chainstate, CEvoDB& evoDb, LLMQContext& llmq_ctx, MNSubsidyAgent& mn_subsidy,
-                               const CTxMemPool& mempool, const CChainParams& params)
-    : BlockAssembler(chainstate, evoDb, llmq_ctx, mn_subsidy, mempool, params, DefaultOptions()) {}
+BlockAssembler::BlockAssembler(CChainState& chainstate, CEvoDB& evoDb, CMNHFManager& mnhfman, LLMQContext& llmq_ctx,
```

Comment:
I don't like constructor of BlockAssembler to accept many-many arguments; this commit make list of arguments longer --- I don't feel like that is a useful refactoring, moreover I feel like that's getting worse. 

If that's a staging temporary changes - I'd like to know what is a next step, but BlockAssembler having +1 more argument as it is done now that is defactoring, not refactoring in my opinion

--------------
IMO much better to let BlockAssembler to use mnhfman, creditpool, from chainparams or from cs_helper rather than have that long list of arguments in constructor

Reply:
we could simply pass node context here I guess c764c09255

Reply:
I don't know how I feel about making the managers available from `cs_helper` (everywhere else we accept the references for the classes' use but don't actually expose it) but I do agree, the long arguments are frustration-inducing (I'd know, reordering commits and cleaning up required me to fix up the argument list manually multiple times).

But the truth of the matter is, we have a lot of managers. Excluding CoinJoin and LLMQ managers (which are managed by their own context), we have ~10 managers and I don't think that's all of them. There's no real way to avoid it without consolidating them. There's a reason I got rid of `CGovernanceTriggerManager`, we don't need more managers.

I did think about creating `DashContext` but then accessing something like `CBLSWorker` would require doing something like `node.dash_ctx->llmq_ctx->bls_worker` and that, at least on initial impression, seems like too much nesting.

---

Using `CChainState` seems unideal imo, I've tried experimenting with aliasing the `CDeterministicMNList` query function to `CChainState` (similar to what was done with `CMNHFManager`) but it seemed unwieldly to do so, not only was I adding yet another argument to pass through `ChainstateManager` and `CChainState`, for a function that `CChainState` doesn't really need, it's more likely than not that if `CDeterministicMNList` is passed upwards enough, we'll eventually reach a point where `CDeterministicMNManager` would be needed.

The reason why I did the refactoring for `CMNHFManager` is because I'd prefer if `CChainState` accepts the fewest possible arguments. It's the reason `ChainstateHelper` exists, because the alternative would be piping `CGovernanceManager`, `CSporkManager` and others.

---

I'm not quite sure of adding `NodeContext` as an argument, it doesn't really seem to be meant to be used that way. Especially since, unlike `LLMQContext`, where the constructor initializes its members, there's no such guarantee for `NodeContext.

Reply:
Cherry-picked and squashed in latest push